x/params/store/store.go

```diff
@@ -0,0 +1,171 @@
+package store
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// Individual parameter store for each keeper
+type Store struct {
+	cdc  *codec.Codec
+	key  sdk.StoreKey
+	tkey sdk.StoreKey
+
+	space []byte
+}
+
+// NewStore constructs a store with namestore
+func NewStore(cdc *codec.Codec, key sdk.StoreKey, tkey sdk.StoreKey, space string) Store {
+	return Store{
+		cdc:  cdc,
+		key:  key,
+		tkey: tkey,
+
+		space: []byte(space),
+	}
+}
+
+// Returns a KVStore identical with ctx,TransientStore(s.key).Prefix()
+func (s Store) kvStore(ctx sdk.Context) sdk.KVStore {
+	return ctx.KVStore(s.key).Prefix(append(s.space, '/'))
+}
+
+// Returns a KVStore identical with ctx.TransientStore(s.tkey).Prefix()
+func (s Store) transientStore(ctx sdk.Context) sdk.KVStore {
+	return ctx.TransientStore(s.tkey).Prefix(append(s.space, '/'))
+}
+
+// Get parameter from store
+func (s Store) Get(ctx sdk.Context, key string, ptr interface{}) {
+	store := s.kvStore(ctx)
+	bz := store.Get([]byte(key))
+	err := s.cdc.UnmarshalJSON(bz, ptr)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// GetIfExists do not modify ptr if the stored parameter is nil
+func (s Store) GetIfExists(ctx sdk.Context, key string, ptr interface{}) {
+	store := s.kvStore(ctx)
+	bz := store.Get([]byte(key))
+	if bz == nil {
+		return
+	}
+	err := s.cdc.UnmarshalJSON(bz, ptr)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Get raw bytes of parameter from store
+func (s Store) GetRaw(ctx sdk.Context, key string) []byte {
+	store := s.kvStore(ctx)
+	res := store.Get([]byte(key))
+	return res
+}
+
+// Check if the parameter is set in the store
+func (s Store) Has(ctx sdk.Context, key string) bool {
+	store := s.kvStore(ctx)
+	return store.Has([]byte(key))
+}
+
+// Returns true if the parameter is set in the block
+func (s Store) Modified(ctx sdk.Context, key string) bool {
+	tstore := s.transientStore(ctx)
+	return tstore.Has([]byte(key))
+}
+
+// Set parameter, return error if stored parameter has different type from input
+func (s Store) Set(ctx sdk.Context, key string, param interface{}) {
+	store := s.kvStore(ctx)
+	keybz := []byte(key)
+
+	bz := store.Get(keybz)
+	if bz != nil {
```

Comment:
Can we add a comment as to why we do this?

Reply:
Huh maybe there's a comment now but I'm still not sure I think this is necessary. An invalid parameter set is an error on the part of the programmer - do we really need to check it at runtime? Calling `reflect` is expensive.

Reply:
If we're sure this is called rarely I guess this is OK, it is a nice sanity check.

Reply:
Also the master module may not know what the parameter is intended for, so there is some possibility that the master module will set wrong value.

Reply:
In that case should we block the master module from setting anything in the subspace?

Reply:
Why are we doing reflection? Can't we just amino decode into a struct already of the correct type?

Reply:
Just had a conversation with mossid about this. We basically ran through the options to guarantee type safety. This seems like the best to me. The other options were pass in a second parameter of expected type (which feels redundant), and to keep a map of key to type. This map idea seems extremely hard to get right in the presence of structs getting renamed, etc. and maintaining cross language support + support across upgrades.

This approach will work between protobuf struct upgrades (AFAICT). Additionally param Set's should be rare, so its probably worth ensuring type safety. (It would be quite annoying to have set a wrong type on chain, and then working out a way to restart from the inevitable failure it would cause once called)

Reply:
I don't think we should be doing this -- regardless of amino using reflection, we should register types (like RegisterConcrete) for param store values (and use reflection to ensure that future gets/sets use the correct type).

> structs getting renamed, etc. and maintaining cross language support + support across upgrades.

Lets just say that you can never change the param type once you set it in the app.go file via a RegisterParamType()?

Reply:
I agree with that, but can we pass a `string=>reflect.Type` table to the constructor of `param.Store` instead of using `RegisterParamType`? It makes the type of the parameters naturally immutable(there cannot be accidental redeclaration) and reduces code size.